common/frame_session.go

```diff
@@ -1187,18 +1197,20 @@ func (fs *FrameSession) updateViewport() error {
 		return fmt.Errorf("emulating viewport: %w", err)
 	}
 
-	// add an inset to viewport depending on the operating system.
-	// this won't add an inset if we're running in headless mode.
-	viewport.calculateInset(
-		fs.page.browserCtx.browser.browserOpts.Headless,
-		runtime.GOOS,
-	)
-	action2 := browser.SetWindowBounds(fs.windowID, &browser.Bounds{
-		Width:  viewport.Width,
-		Height: viewport.Height,
-	})
-	if err := action2.Do(cdp.WithExecutor(fs.ctx, fs.session)); err != nil {
-		return fmt.Errorf("setting window bounds: %w", err)
+	if fs.windowID != nil {
```

Comment:
Instead of using a pointer, why don't we just use `hasUIWindow` parameter here?

Reply:
Not sure what you mean. As in, save the flag as a field of the `FrameSession` type and then use it here instead of windowID being a pointer?

Reply:
Ah, sorry, I wasn't clear. Yep, instead of using a pointer (which can be race-condition-prone and make an allocation for each frame), adding a `hasUIWindow` field is less dangerous and also communicates better that the current frame has no window UI. Also, we can use it for other purposes in the future if needed. Besides, `windowID` will be zero when it's absent since the Chromium code uses a `nullptr`, which evaluates to zero.

Reply:
I wanted to test what happens when `windowID` was `0`, just incase we could do something a little more clever and to get a better understanding of `nullptr`, but it turns out that chromium must interpret it as non null (or the error is incorrect from chromium) as it returns `Browser window not found`. In conclusion I think your suggestion still stands (`hasUIWindow` as a field on `FrameSession`) as the safest path forward 🙂

Reply:
Updated in ed17d77f74c91ce7905ad3ba326cf01b8bac573c

Reply:
Thanks for giving this a try 🙇 :)